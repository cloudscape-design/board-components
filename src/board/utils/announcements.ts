// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { MIN_COL_SPAN, MIN_ROW_SPAN } from "../../internal/constants";
import { Direction, ItemId } from "../../internal/interfaces";
import {
  BoardProps,
  DndActionAnnouncement,
  ItemRemovedAnnouncement,
  Transition,
  TransitionAnnouncement,
} from "../interfaces";

/**
 * Creates an announcement object describing the last user move.
 */
export function createOperationAnnouncement<D>(
  transition: Transition<D>,
  direction: null | Direction
): null | DndActionAnnouncement {
  const { operation, layoutShift, itemsLayout } = transition;
  const targetItem = itemsLayout.items.find((it) => it.id === transition.draggableItem.id) ?? null;

  if (!layoutShift) {
    return null;
  }

  const firstMove = layoutShift.moves[0];
  const targetId = firstMove?.itemId ?? targetItem?.id;
  if (!targetId) {
    return null;
  }

  const itemMoves = layoutShift.moves.filter((m) => m.itemId === targetId);
  const lastItemMove = itemMoves[itemMoves.length - 1];
  const placement = lastItemMove ?? targetItem;

  const conflicts = new Set(layoutShift.conflicts);

  const disturbed = new Set(layoutShift.moves.map((move) => move.itemId));
  disturbed.delete(targetId);

  return {
    type: "dnd-action",
    item: transition.draggableItem,
    operation,
    placement: {
      x: placement.x,
      y: placement.y,
      width: placement.width,
      height: placement.height,
    },
    direction,
    conflicts,
    disturbed,
  };
}

/**
 * Applies i18nStrings to the announcement object to produce a string for the live region.
 */
export function announcementToString<D>(
  announcement: TransitionAnnouncement,
  items: readonly BoardProps.Item<D>[],
  i18nStrings: BoardProps.I18nStrings<D>
): string {
  if (!announcement) {
    return "";
  }
  const item = announcement.item as BoardProps.Item<D>;

  const toItem = (id: ItemId) => items.find((it) => it?.id === id)!;
  const formatDirection = (direction: null | Direction) => {
    if (!direction) {
      return null;
    }
    return direction === "left" || direction === "right" ? "horizontal" : "vertical";
  };

  function createDndActionAnnouncement(announcement: DndActionAnnouncement) {
    const placement = announcement.placement;
    const direction = formatDirection(announcement.direction);
    const conflicts = [...announcement.conflicts].map(toItem);
    const disturbed = [...announcement.disturbed].map(toItem);

    switch (announcement.operation) {
      case "reorder":
        return i18nStrings.liveAnnouncementDndItemReordered({
          item,
          placement,
          direction: direction!,
          conflicts,
          disturbed,
        });
      case "resize":
        return i18nStrings.liveAnnouncementDndItemResized({
          item,
          placement,
          direction: direction!,
          isMinimalColumnsReached: placement.width === (item.definition?.minColumnSpan ?? MIN_COL_SPAN),
          isMinimalRowsReached: placement.height === (item.definition?.minRowSpan ?? MIN_ROW_SPAN),
          conflicts,
          disturbed,
        });
      case "insert":
        return i18nStrings.liveAnnouncementDndItemInserted({ item, placement, conflicts, disturbed });
    }
  }

  function createItemRemovedAnnouncement(announcement: ItemRemovedAnnouncement) {
    return i18nStrings.liveAnnouncementItemRemoved({ item, disturbed: [...announcement.disturbed].map(toItem) });
  }

  switch (announcement.type) {
    case "dnd-started":
      return i18nStrings.liveAnnouncementDndStarted(announcement.operation);
    case "dnd-action":
      return createDndActionAnnouncement(announcement);
    case "dnd-committed":
      return i18nStrings.liveAnnouncementDndCommitted(announcement.operation);
    case "dnd-discarded":
      return i18nStrings.liveAnnouncementDndDiscarded(announcement.operation);
    case "item-removed":
      return createItemRemovedAnnouncement(announcement);
  }
}
