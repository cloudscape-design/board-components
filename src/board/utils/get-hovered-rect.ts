// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { GridLayoutItem, ItemId } from "../../internal/interfaces";

/**
 * Creates a minimal hovered rectangle (in grid units) that contains all collided placeholders.
 */
export function getHoveredRect(collisionsIds: readonly ItemId[], placeholders: readonly GridLayoutItem[]) {
  const hoveredPlaceholders = collisionsIds.map((id) => placeholders.find((p) => p.id === id)!);
  return hoveredPlaceholders.reduce(
    (rect, collision) => ({
      top: Math.min(rect.top, collision.y),
      left: Math.min(rect.left, collision.x),
      bottom: Math.max(rect.bottom, collision.y + collision.height),
      right: Math.max(rect.right, collision.x + collision.width),
    }),
    {
      top: Number.POSITIVE_INFINITY,
      left: Number.POSITIVE_INFINITY,
      bottom: Number.NEGATIVE_INFINITY,
      right: Number.NEGATIVE_INFINITY,
    }
  );
}
