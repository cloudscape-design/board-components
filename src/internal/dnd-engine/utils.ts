// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Item, Position } from "./interfaces";

export interface Rect {
  left: number;
  right: number;
  top: number;
  bottom: number;
}

export function sortGridItems(items: readonly Item[]): readonly Item[] {
  return [...items].sort((a, b) => (b.y - a.y === 0 ? b.x - a.x : b.y - a.y));
}

export function getItemRect(item: Item): Rect {
  return {
    left: item.x,
    right: item.x + item.width - 1,
    top: item.y,
    bottom: item.y + item.height - 1,
  };
}

export function normalizePath(origin: Position, path: readonly Position[]): readonly Position[] {
  // Remove path prefixes that return to the original location.
  for (let i = 0; i < path.length; i++) {
    if (path[i].x === origin.x && path[i].y === origin.y) {
      path = path.slice(i + 1);
    }
  }

  // Store last visited indexes per position.
  const positionToLastIndex = new Map<string, number>();
  for (let index = 0; index < path.length; index++) {
    positionToLastIndex.set(`${path[index].x}:${path[index].y}`, index);
  }

  // Compose path from last visited indixes only.
  const normalizedPath: Position[] = [];
  let index = 0;
  while (index < path.length) {
    const lastVisitedIndex = positionToLastIndex.get(`${path[index].x}:${path[index].y}`)!;
    normalizedPath.push(path[lastVisitedIndex]);
    index = lastVisitedIndex + 1;
  }

  return normalizedPath;
}
