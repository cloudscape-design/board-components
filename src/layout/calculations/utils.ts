// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { GridLayoutItem } from "../../internal/base-types";
import { Rect } from "./interfaces";

export function itemToRect(item: GridLayoutItem): Rect {
  return {
    left: item.x,
    top: item.y,
    right: item.x + item.width,
    bottom: item.y + item.height,
  };
}

export function isInsideRect(rect: Rect, bounds: Rect) {
  return (
    rect.top <= bounds.top && rect.left >= bounds.left && rect.right <= bounds.right && rect.bottom >= bounds.bottom
  );
}
