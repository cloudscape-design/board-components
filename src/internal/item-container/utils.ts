// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { getLogicalBoundingClientRect, getLogicalClientX } from "@cloudscape-design/component-toolkit/internal";
import { InteractionState } from "@cloudscape-design/components/internal/components/drag-handle/hooks/interfaces";

import { Operation } from "../dnd-controller/controller"; // Adjust this path
import { Coordinates } from "../utils/coordinates"; // Adjust this path based on your project structure
import { CLICK_DRAG_THRESHOLD, Transition } from ".";

export type HandleActiveState = null | "pointer" | "uap";

export interface DetermineHandleActiveStateArgs {
  isHandleActive: boolean;
  currentTransition: Transition | null;
  interactionHookValue: InteractionState;
  targetOperation: Operation;
}

export function getDndOperationType(uiOperation: "drag" | "resize", isItemPlaced: boolean): Operation {
  if (uiOperation === "resize") {
    return "resize";
  }
  return isItemPlaced ? "reorder" : "insert";
}

interface CalculateInitialPointerDataArgs {
  event: PointerEvent;
  operation: "drag" | "resize";
  rect: ReturnType<typeof getLogicalBoundingClientRect>;
  getMinSize: () => { minWidth: number; minHeight: number };
  isRtl: boolean;
}

/**
 * Calculates the initial pointer offset and boundaries for a drag or resize interaction
 * to help determine how the item's movement or resizing behaves relative to the pointer.
 */
export function calculateInitialPointerData({
  event,
  operation,
  rect,
  getMinSize,
  isRtl,
}: CalculateInitialPointerDataArgs): {
  pointerOffset: Coordinates;
  pointerBoundaries: Coordinates | null;
} {
  const clientX = getLogicalClientX(event, isRtl);
  const clientY = event.clientY;

  let pointerOffset: Coordinates;
  let pointerBoundaries: Coordinates | null = null;

  if (operation === "resize") {
    // For resize, offset is calculated from the bottom-right corner.
    pointerOffset = new Coordinates({
      x: clientX - rect.insetInlineEnd,
      y: clientY - rect.insetBlockEnd,
    });

    // Boundaries to ensure resize doesn't go below minimum dimensions.
    const { minWidth, minHeight } = getMinSize();
    pointerBoundaries = new Coordinates({
      x: clientX - rect.inlineSize + minWidth,
      y: clientY - rect.blockSize + minHeight,
    });
  } else {
    // For drag, offset is calculated from the top-left corner.
    pointerOffset = new Coordinates({
      x: clientX - rect.insetInlineStart,
      y: clientY - rect.insetBlockStart,
    });
  }

  return { pointerOffset, pointerBoundaries };
}

export function hasPointerMovedBeyondThreshold(
  event: PointerEvent,
  initialPosition: { x: number; y: number } | undefined,
  threshold: number = CLICK_DRAG_THRESHOLD,
): boolean {
  if (!initialPosition) {
    return false;
  }
  return (
    event.clientX > initialPosition.x + threshold ||
    event.clientX < initialPosition.x - threshold ||
    event.clientY > initialPosition.y + threshold ||
    event.clientY < initialPosition.y - threshold
  );
}

export function determineHandleActiveState({
  isHandleActive,
  currentTransition,
  interactionHookValue,
  targetOperation,
}: DetermineHandleActiveStateArgs): HandleActiveState {
  if (isHandleActive && currentTransition?.operation === targetOperation && interactionHookValue === "dnd-start") {
    return "pointer";
  } else if (
    isHandleActive &&
    currentTransition?.operation === targetOperation &&
    interactionHookValue === "uap-action-start"
  ) {
    return "uap";
  }
  return null;
}
