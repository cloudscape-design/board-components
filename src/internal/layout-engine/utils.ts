// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { GridLayoutItem } from "../interfaces";
import { Position } from "../utils/position";

export interface Rect {
  left: number;
  right: number;
  top: number;
  bottom: number;
}

export function sortGridItems(items: readonly GridLayoutItem[]): readonly GridLayoutItem[] {
  return [...items].sort((a, b) => (b.y - a.y === 0 ? b.x - a.x : b.y - a.y));
}

export function getItemRect(item: GridLayoutItem): Rect {
  return {
    left: item.x,
    right: item.x + item.width - 1,
    top: item.y,
    bottom: item.y + item.height - 1,
  };
}

export function normalizeMovePath(origin: Position, path: readonly Position[]): readonly Position[] {
  path = normalizePathOrigin(origin, path);

  // Store last visited indexes per position.
  const positionToLastIndex = new Map<string, number>();
  for (let index = 0; index < path.length; index++) {
    positionToLastIndex.set(`${path[index].x}:${path[index].y}`, index);
  }

  // Compose path from last visited indices only.
  const normalizedPath: Position[] = [];
  let index = 0;
  while (index < path.length) {
    const lastVisitedIndex = positionToLastIndex.get(`${path[index].x}:${path[index].y}`)!;
    normalizedPath.push(path[lastVisitedIndex]);
    index = lastVisitedIndex + 1;
  }

  return normalizePathSteps(origin, normalizedPath);
}

export function normalizeResizePath(origin: Position, path: readonly Position[]): readonly Position[] {
  path = normalizePathOrigin(origin, path);

  if (path.length === 0) {
    return [];
  }

  const normalizedPath: Position[] = [path[path.length - 1]];
  for (let stepIndex = path.length - 2; stepIndex >= 0; stepIndex--) {
    const prev = normalizedPath[normalizedPath.length - 1];
    const current = path[stepIndex];
    if (current.x < prev.x || current.y < prev.y) {
      normalizedPath.push(current);
    }
  }
  normalizedPath.reverse();

  return normalizePathSteps(origin, normalizedPath);
}

// Removes path prefixes that return to the original location.
function normalizePathOrigin(origin: Position, path: readonly Position[]): readonly Position[] {
  let lastOriginIndex = -1;
  for (let i = 0; i < path.length; i++) {
    if (path[i].x === origin.x && path[i].y === origin.y) {
      lastOriginIndex = i;
    }
  }
  return path.slice(lastOriginIndex + 1);
}

// Ensures path only includes single-length steps.
function normalizePathSteps(origin: Position, path: readonly Position[]): readonly Position[] {
  const normalizedPath: Position[] = [];

  let prevX = origin.x;
  let prevY = origin.y;

  for (const step of path) {
    const vx = Math.sign(step.x - prevX);
    const vy = Math.sign(step.y - prevY);

    for (let x = prevX, y = prevY; x !== step.x || y !== step.y; ) {
      if (x !== step.x) {
        x += vx;
      } else {
        y += vy;
      }
      normalizedPath.push(new Position({ x, y }));
    }

    prevX = step.x;
    prevY = step.y;
  }

  return normalizedPath;
}
