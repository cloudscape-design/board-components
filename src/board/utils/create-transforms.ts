// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { GridContext } from "../../internal/grid/interfaces";
import { GridLayout, ItemId, Transform } from "../../internal/interfaces";
import { CommittedMove } from "../../internal/layout-engine/interfaces";

/**
 * Creates a mapping of between items and transforms to be applied while in transition.
 */
export function createTransforms(grid: GridLayout, moves: readonly CommittedMove[], gridContext: GridContext) {
  const transforms: Record<ItemId, Transform> = {};

  for (const move of moves) {
    const item = grid.items.find((prev) => prev.id === move.itemId);

    if (move.type === "REMOVE") {
      transforms[move.itemId] = { type: "remove" };
    } else if (item) {
      transforms[item.id] = {
        type: "move",
        x: gridContext.getColOffset(move.x - item.x),
        y: gridContext.getRowOffset(move.y - item.y),
        width: gridContext.getWidth(move.width),
        height: gridContext.getHeight(move.height),
      };
    }
  }

  return transforms;
}
