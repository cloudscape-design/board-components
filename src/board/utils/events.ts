// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { LayoutShift } from "../../internal/layout-engine/interfaces";
import { createCustomEvent } from "../../internal/utils/events";
import { transformItems } from "../../internal/utils/layout";
import { BoardProps } from "../interfaces";

export function createItemsChangeEvent<D>(
  items: readonly BoardProps.Item<D>[],
  layoutShift: LayoutShift
): CustomEvent<BoardProps.ItemsChangeDetail<D>> {
  const insertTarget = layoutShift.moves.find((move) => move.type === "INSERT")?.itemId ?? null;
  const removeTarget = layoutShift.moves.find((move) => move.type === "REMOVE")?.itemId ?? null;
  const resizeTarget = layoutShift.moves.find((move) => move.type === "RESIZE")?.itemId ?? null;

  const newItems = transformItems(items, layoutShift.next, resizeTarget ?? insertTarget);
  const addedItem = newItems.find((it) => it.id === insertTarget);
  const removedItem = newItems.find((it) => it.id === removeTarget);

  return createCustomEvent({ items: newItems, addedItem, removedItem });
}
