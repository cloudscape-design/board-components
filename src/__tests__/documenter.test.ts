// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { expect, test } from "vitest";
import { getAllComponents, requireComponentDefinition } from "./utils";

test.each<string>(getAllComponents())(`definition for %s matches the snapshot`, (componentName: string) => {
  const definition = requireComponentDefinition(componentName);
  expect(definition).toMatchSnapshot(componentName);
});
