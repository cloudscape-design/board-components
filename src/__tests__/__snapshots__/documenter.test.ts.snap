// Vitest Snapshot v1

exports[`definition for board matches the snapshot > board 1`] = `
{
  "events": [
    {
      "cancelable": false,
      "description": "Fired when a user interaction changes size or position of board items.",
      "detailInlineType": {
        "name": "BoardProps.ItemsChangeDetail",
        "properties": [
          {
            "name": "addedItem",
            "optional": true,
            "type": "BoardProps.Item<D>",
          },
          {
            "name": "items",
            "optional": false,
            "type": "ReadonlyArray<BoardProps.Item<D>>",
          },
          {
            "name": "removedItem",
            "optional": true,
            "type": "BoardProps.Item<D>",
          },
        ],
        "type": "object",
      },
      "detailType": "BoardProps.ItemsChangeDetail<D>",
      "name": "onItemsChange",
    },
  ],
  "functions": [],
  "name": "Board",
  "properties": [
    {
      "description": "An object containing all the necessary localized strings required by the component.",
      "inlineType": {
        "name": "BoardProps.I18nStrings",
        "properties": [
          {
            "name": "liveAnnouncementOperation",
            "optional": false,
            "type": "(operation: BoardProps.OperationState<D>) => string",
          },
          {
            "name": "liveAnnouncementOperationCommitted",
            "optional": false,
            "type": "(operationType: BoardProps.DragOperationType) => string",
          },
          {
            "name": "liveAnnouncementOperationDiscarded",
            "optional": false,
            "type": "(operationType: BoardProps.DragOperationType) => string",
          },
          {
            "name": "liveAnnouncementOperationStarted",
            "optional": false,
            "type": "(operationType: BoardProps.OperationType) => string",
          },
          {
            "name": "navigationAriaDescription",
            "optional": true,
            "type": "string",
          },
          {
            "name": "navigationAriaLabel",
            "optional": false,
            "type": "string",
          },
          {
            "name": "navigationItemAriaLabel",
            "optional": false,
            "type": "(item: BoardProps.Item<D> | null) => string",
          },
        ],
        "type": "object",
      },
      "name": "i18nStrings",
      "optional": false,
      "type": "BoardProps.I18nStrings<D>",
    },
    {
      "description": "Specifies the items displayed in the board. Each item is includes its position on the board and
attached data. The content of an item is controlled by the \`renderItem\` property.",
      "name": "items",
      "optional": false,
      "type": "ReadonlyArray<BoardProps.Item<D>>",
    },
    {
      "description": "Specifies a function to render a board item content. The return value must include board item component.",
      "name": "renderItem",
      "optional": false,
      "type": "(item: BoardProps.Item<D>, actions: BoardProps.ItemActions) => Element",
    },
  ],
  "regions": [
    {
      "description": "Rendered when the \`items\` array is empty.
When items are loading the slot can be used to render the loading indicator.
",
      "isDefault": false,
      "name": "empty",
    },
  ],
  "releaseStatus": "stable",
}
`;

exports[`definition for board-item matches the snapshot > board-item 1`] = `
{
  "events": [],
  "functions": [],
  "name": "BoardItem",
  "properties": [
    {
      "description": "Determines whether the main content of the item has padding. If \`true\`, removes the default padding
from the content area.",
      "name": "disableContentPaddings",
      "optional": true,
      "type": "boolean",
    },
    {
      "description": "An object containing all the necessary localized strings required by the component.",
      "inlineType": {
        "name": "BoardItemProps.I18nStrings",
        "properties": [
          {
            "name": "dragHandleAriaDescription",
            "optional": true,
            "type": "string",
          },
          {
            "name": "dragHandleAriaLabel",
            "optional": false,
            "type": "string",
          },
          {
            "name": "resizeHandleAriaDescription",
            "optional": true,
            "type": "string",
          },
          {
            "name": "resizeHandleAriaLabel",
            "optional": false,
            "type": "string",
          },
        ],
        "type": "object",
      },
      "name": "i18nStrings",
      "optional": false,
      "type": "BoardItemProps.I18nStrings",
    },
  ],
  "regions": [
    {
      "description": "Main content of the item.",
      "isDefault": true,
      "name": "children",
    },
    {
      "description": "Footer of the item.",
      "isDefault": false,
      "name": "footer",
    },
    {
      "description": "Heading element of the item. Use the [header component](/components/header/).",
      "isDefault": false,
      "name": "header",
    },
    {
      "description": "Additional slot next to the heading. Use it to render an overflow actions menu in the form of a button dropdown.",
      "isDefault": false,
      "name": "settings",
    },
  ],
  "releaseStatus": "stable",
}
`;

exports[`definition for items-palette matches the snapshot > items-palette 1`] = `
{
  "events": [],
  "functions": [],
  "name": "ItemsPalette",
  "properties": [
    {
      "description": "An object containing all the necessary localized strings required by the component.",
      "inlineType": {
        "name": "ItemsPaletteProps.I18nStrings",
        "properties": [
          {
            "name": "liveAnnouncementDragDiscarded",
            "optional": false,
            "type": "string",
          },
          {
            "name": "liveAnnouncementDragStarted",
            "optional": false,
            "type": "string",
          },
          {
            "name": "navigationAriaDescription",
            "optional": true,
            "type": "string",
          },
          {
            "name": "navigationAriaLabel",
            "optional": false,
            "type": "string",
          },
          {
            "name": "navigationItemAriaLabel",
            "optional": false,
            "type": "(item: ItemsPaletteProps.Item<D>) => string",
          },
        ],
        "type": "object",
      },
      "name": "i18nStrings",
      "optional": false,
      "type": "ItemsPaletteProps.I18nStrings<D>",
    },
    {
      "description": "Specifies the items displayed in the palette. The content of each item is controlled by the \`renderItem\` property.",
      "name": "items",
      "optional": false,
      "type": "ReadonlyArray<ItemsPaletteProps.Item<D>>",
    },
    {
      "description": "Specifies a function to render a palette item content. The return value must include board item component.",
      "name": "renderItem",
      "optional": false,
      "type": "(item: ItemsPaletteProps.Item<D>, context: ItemsPaletteProps.ItemContext) => Element",
    },
  ],
  "regions": [],
  "releaseStatus": "stable",
}
`;
